titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultDirectKubeApiServerIntegrator.java

```diff
@@ -0,0 +1,183 @@
+import com.netflix.titus.common.util.guice.annotation.Activator;
+import com.netflix.titus.master.mesos.kubeapiserver.direct.model.PodEvent;
+import io.kubernetes.client.ApiClient;
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.apis.CoreV1Api;
+import io.kubernetes.client.informer.ResourceEventHandler;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.informer.SharedInformerFactory;
+import io.kubernetes.client.models.V1Pod;
+import io.kubernetes.client.models.V1PodList;
+import io.kubernetes.client.util.CallGeneratorParams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import static com.netflix.titus.master.mesos.kubeapiserver.KubeUtil.createSharedInformerFactory;
+
+@Singleton
+public class DefaultDirectKubeApiServerIntegrator implements DirectKubeApiServerIntegrator {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultDirectKubeApiServerIntegrator.class);
+
+    private static final String KUBERNETES_NAMESPACE = "default";
+
+    private final DirectKubeConfiguration configuration;
+
+    private final ApiClient apiClient;
+    private final CoreV1Api coreV1Api;
+    private final TaskToPodConverter taskToPodConverter;
+
+    private volatile SharedInformerFactory sharedInformerFactory;
+    private volatile SharedIndexInformer<V1Pod> podInformer;
+
+    private final ConcurrentMap<String, V1Pod> pods = new ConcurrentHashMap<>();
+
+    @Inject
+    public DefaultDirectKubeApiServerIntegrator(DirectKubeConfiguration configuration,
+                                                ApiClient apiClient,
+                                                TaskToPodConverter taskToPodConverter) {
+        this.configuration = configuration;
+        this.apiClient = apiClient;
+        this.coreV1Api = new CoreV1Api(apiClient);
+        this.taskToPodConverter = taskToPodConverter;
+    }
+
+    @Activator
+    public void enterActiveMode() {
+        this.sharedInformerFactory = createSharedInformerFactory(
+                "kube-api-server-integrator-shared-informer-",
+                apiClient
+        );
+        this.podInformer = createPodInformer(sharedInformerFactory, coreV1Api);
+        sharedInformerFactory.startAllRegisteredInformers();
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        if (sharedInformerFactory != null) {
+            sharedInformerFactory.stopAllRegisteredInformers();
+        }
+    }
+
+    @Override
+    public Map<String, V1Pod> getPods() {
+        return new HashMap<>(pods);
+    }
+
+    @Override
+    public Mono<V1Pod> launchTask(Job job, Task task) {
+        return Mono.defer(() -> {
+            try {
+                V1Pod v1Pod = taskToPodConverter.apply(job, task);
+                logger.info("creating pod: {}", v1Pod);
+                coreV1Api.createNamespacedPod(KUBERNETES_NAMESPACE, v1Pod, null, null, null);
```

Comment:
this will block the Reactor thread, consider using the async version, or forcing it to be on `Schedulers#elastic()` or our own `ExecutorService` thread pool